File: CellsToServersCpp/HierarchicalTilingAppCpp/SumOfSquares/BacktrackingMethod.cpp
#include "BacktrackingMethod.hpp"

namespace SumOfSquares
{
	BacktrackingMethod::BacktrackingMethod(CornacchiaMethod& cornacchiaMethod) : 
		cornacchiaMethod(cornacchiaMethod)
	{
	}

	Vector_t& BacktrackingMethod::decomposeByBacktracking(int num, int squaresNO)
	{
		Vector_t * decompositions = new Vector_t();
		if (squaresNO == 1)
		{
			int floorOfSquareOfN = (int)floor(sqrt(num));
			if (floorOfSquareOfN * floorOfSquareOfN == num)
			{
				int * newTuple = new int[1]; newTuple[0] = floorOfSquareOfN;
				IntTuple * newIntTuple = new IntTuple(1, newTuple);
				decompositions->push_back(newIntTuple);
			}
		}
		else if (squaresNO >= 2)
		{
			decompositions = innerDecomposeByBacktracking(num, squaresNO);
		}
		return *decompositions;
	}

	Vector_t * BacktrackingMethod::innerDecomposeByBacktracking(int num, int squaresNO)
	{
		Vector_t * intTuples = new Vector_t();
		if (num == 0)
		{
			int * newTuple = new int[squaresNO];
			for (int idx = 0; idx < squaresNO; idx++)
			{
				newTuple[idx] = 0;
			}
			IntTuple * newIntTuple = new IntTuple(squaresNO, newTuple);
			intTuples->push_back(newIntTuple);
		} 
		else if (squaresNO == 2)
			intTuples = &(cornacchiaMethod.applyCornacchiaMethod(num));
		else
		{
			intTuples = innerDecomposeByBacktrackingWhenSquaresNOIsGreaterThanTwo(num, squaresNO);
		}
		return intTuples;
	}

	Vector_t * BacktrackingMethod::innerDecomposeByBacktrackingWhenSquaresNOIsGreaterThanTwo(int num,
		int squaresNO)
	{
		Vector_t * intTuples = new Vector_t();
		List_t * tempIntTuples = new List_t();
		int floorOfSquareOfN = (int)floor(sqrt(num));
		// This method return such int tuple where the components of this tuple has ascending order.
		// Thus num-lastTerm*lastTerm <= (squaresNO-1)*(lastTerm*lastTerm) 
		// --> sqrt(num/squaresNO) <= lastTerm.
		// (E.g. num = x^2 + y^2 + z^2 (x<=y<=z), num - z^2 = x^2 + y^2 <= 2 * z^2 ...)
		int lowerLimit = (int)(floorOfSquareOfN / sqrt(squaresNO));
		for (int lastTerm = floorOfSquareOfN; lastTerm >= lowerLimit; lastTerm--)
		{
			Vector_t * subIntTuples = innerDecomposeByBacktracking(num - lastTerm * lastTerm,
				squaresNO - 1);
			processSubIntTuples(subIntTuples, lastTerm, squaresNO, tempIntTuples);
		}
		for(List_t::iterator itr = tempIntTuples->begin(); itr != tempIntTuples->end(); itr++)
		{
			intTuples->push_back(*itr);
		}
		return intTuples;
	}

	void BacktrackingMethod::processSubIntTuples(Vector_t * subIntTuples, int lastTerm, int squaresNO,
		List_t * tempIntTuples)
	{
		for(Vector_t::iterator itr = subIntTuples->begin(); itr != subIntTuples->end(); itr++)
		{
			IntTuple * subIntTuple = *itr;
			int * currentTuple = subIntTuple->getTuple();
			// the components of this tuple must have ascending order
			if (currentTuple[squaresNO - 2] <= lastTerm)
			{
				int * newTuple = new int[squaresNO];
				for (int idx = 0; idx < squaresNO; idx++)
				{
					newTuple[idx] = currentTuple[idx];
				}
				newTuple[squaresNO - 1] = lastTerm;
				IntTuple * newIntTuple = new IntTuple(squaresNO, newTuple);
				tempIntTuples->push_back(newIntTuple);
			}
		}
	}
}

File: MeasureApp/MeasureApp/SumOfSquares/BacktrackingMethod.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MeasureApp.SumOfSquares
{
    /// <summary>
    /// This class is based on Peter Schorn's (Institut für Theoretische Informatik, ETH Zürich) technical report
    /// links:  ftp://ftp.inf.ethz.ch/pub/software/xyz/papers/SquareDecomposition.ps
    ///         http://schorn.ch/lagrange.html
    /// </summary>
    public class BacktrackingMethod
    {
        private CornacchiaMethod cornacchiaMethod;

        public BacktrackingMethod(CornacchiaMethod cornacchiaMethod)
        {
            this.cornacchiaMethod = cornacchiaMethod;
        }

        public IntTupleEqualityComparer getComparer()
        {
            return cornacchiaMethod.getComparer();
        }

        /// <summary>
        /// This method return such int tuple where the components of this tuple have ascending order.
        /// </summary>
        /// <param name="num"></param>
        /// <param name="squaresNO"></param>
        /// <returns></returns>
        public IntTuple[] decomposeByBacktracking(int num, int squaresNO)
        {
            IntTuple[] decompositions;
            if (squaresNO == 1)
            {
                int floorOfSquareOfN = (int)Math.Floor(Math.Sqrt(num));
                if (floorOfSquareOfN * floorOfSquareOfN == num)
                {
                    decompositions = new IntTuple[1];
                    decompositions[0] = new IntTuple() { Tuple = new int[] {floorOfSquareOfN} };
                }
                else
                    decompositions = new IntTuple[0];
            }
            else if (squaresNO >= 2)
            {
                decompositions = innerDecomposeByBacktracking(num, squaresNO);
            }
            else
                decompositions = new IntTuple[0];
            return decompositions;
        }

        private IntTuple[] innerDecomposeByBacktracking(int num, int squaresNO)
        {
            IntTuple[] intTuples;
            if (num == 0)
            {
                int[] tuple = new int[squaresNO];
                for (int idx = 0; idx < squaresNO; idx++)
                {
                    tuple[idx] = 0;
                }
                intTuples = new IntTuple[1] { new IntTuple() { Tuple = tuple } };
            } 
            else if (squaresNO == 2)
                intTuples = cornacchiaMethod.applyCornacchiaMethod(num);
            else
            {
                intTuples = innerDecomposeByBacktrackingWhenSquaresNOIsGreaterThanTwo(num, squaresNO);
            }
            return intTuples;
        }

        private IntTuple[] innerDecomposeByBacktrackingWhenSquaresNOIsGreaterThanTwo(int num, int squaresNO)
        {
            IntTuple[] intTuples;
            List<IntTuple> tempIntTuples = new List<IntTuple>();
            int floorOfSquareOfN = (int)Math.Floor(Math.Sqrt(num));
            // This method return such int tuple where the components of this tuple has ascending order.
            // Thus num-lastTerm*lastTerm <= (squaresNO-1)*(lastTerm*lastTerm) --> sqrt(num/squaresNO) <= lastTerm.
            // (E.g. num = x^2 + y^2 + z^2 (x<=y<=z), num - z^2 = x^2 + y^2 <= 2 * z^2 ...)
            int lowerLimit = (int)(floorOfSquareOfN / Math.Sqrt(squaresNO));
            for (int lastTerm = floorOfSquareOfN; lastTerm >= lowerLimit; lastTerm--)
            {
                IntTuple[] subIntTuples = innerDecomposeByBacktracking(num - lastTerm * lastTerm, squaresNO - 1);
                foreach (var subIntTuple in subIntTuples)
                {
                    // the components of this tuple must have ascending order
                    if (subIntTuple.Tuple[squaresNO - 2] <= lastTerm)
                    {
                        int[] tuple = new int[squaresNO];
                        subIntTuple.Tuple.CopyTo(tuple, 0);
                        tuple[squaresNO - 1] = lastTerm;
                        tempIntTuples.Add(new IntTuple() { Tuple = tuple });
                    }
                }
            }
            intTuples = tempIntTuples.ToArray();
            return intTuples;
        }
    }
}
